### PR TITLE
Wrapping of legacy device automatically in various device creation/qnode/execute functions

### PR DESCRIPTION
**Pull request by Shiro-Raven**
**Context:**
With the `LegacyDeviceFacade` now in place, we can add automatic wrapping of legacy devices.

**Description of the Change:**
Add automatic wrapping to `qml.device`, `qml.execute`, `QNode` constructor, and the `get_best_method` and `best_method_str` functions of the QNode class. The tests are also updated accordingly.

**Benefits:**
Users no longer need to worry about upgrading their devices to the new Device API and can use the facade to access the basic functions of the new API.

**Possible Drawbacks:**
The facade doesn't yet provide all potential advantages of fully upgrading to the new API

[[sc-65998](https://app.shortcut.com/xanaduai/story/65998)]

### CODE REVIEW

File: pennylane/workflow/qnode.py [hunk 1]
```diff
@@ -985,9 +805,9 @@ def construct(self, args, kwargs):  # pylint: disable=too-many-branches
         old_interface = self.interface
 
         if self._qfunc_uses_shots_arg:
-            shots = _get_device_shots(self._original_device)
+            shots = self._original_device.shots
```
Comment (albi3ro) [inline pennylane/workflow/qnode.py:808]: Since no more device switching in the qnode, we can get rid of `QNode._original_device` and just use `self.device` everywhere.

File: pennylane/workflow/qnode.py [hunk 2]
```diff
@@ -1080,22 +897,29 @@ def _execution_component(self, args: tuple, kwargs: dict, override_shots) -> qml
 
         execute_kwargs = copy.copy(self.execute_kwargs)
         mcm_config = copy.copy(execute_kwargs["mcm_config"])
-        finite_shots = _get_device_shots(self.device) if override_shots is False else override_shots
+        finite_shots = (lambda dev: dev.shots) if override_shots is False else override_shots
+
         if not finite_shots:
```
Comment (albi3ro) [inline pennylane/workflow/qnode.py:902]:
```suggestion
        if not self._tape.shots:
```

File: pennylane/workflow/qnode.py [hunk 3]
```diff
@@ -1080,22 +897,29 @@ def _execution_component(self, args: tuple, kwargs: dict, override_shots) -> qml
...
                 raise ValueError(
                     f"Cannot use the '{mcm_config.mcm_method}' method for mid-circuit measurements with analytic mode."
                 )
+
         if mcm_config.mcm_method == "single-branch-statistics":
             raise ValueError("Cannot use mcm_method='single-branch-statistics' without qml.qjit.")
 
         full_transform_program = qml.transforms.core.TransformProgram(self.transform_program)
         inner_transform_program = qml.transforms.core.TransformProgram()
         config = None
```
Comment (albi3ro) [inline pennylane/workflow/qnode.py:914]:
```suggestion
```

File: pennylane/workflow/qnode.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/workflow/qnode.py:902]: We might not even need `overrride_shots` in this method anymore.

File: pennylane/workflow/qnode.py [hunk 4]
```diff
@@ -1080,22 +897,29 @@ def _execution_component(self, args: tuple, kwargs: dict, override_shots) -> qml
...
 
-        if isinstance(self.device, qml.devices.Device):
+        config = _make_execution_config(self, self.gradient_fn)
+        device_transform_program, config = self.device.preprocess(execution_config=config)
 
+        if config.use_device_gradient:
+            full_transform_program += device_transform_program
+        else:
+            inner_transform_program += device_transform_program
```
Comment (albi3ro) [inline pennylane/workflow/qnode.py:922]: I might be missing something, but it looks like lines 923-929 need to be deleted.

File: pennylane/workflow/qnode.py [hunk 5]
```diff
@@ -1176,7 +1000,7 @@ def _impl_call(self, *args, **kwargs) -> qml.typing.Result:
             override_shots = False
         else:
             if "shots" not in kwargs:
-                kwargs["shots"] = _get_device_shots(self._original_device)
+                kwargs["shots"] = self._original_device.shots
```
Comment (albi3ro) [inline pennylane/workflow/qnode.py:1003]:
```suggestion
                kwargs["shots"] = self.device.shots

```
No need to use `_original_device` anymore.

File: pennylane/devices/legacy_facade.py [hunk 6]
```diff
@@ -139,12 +139,19 @@ class LegacyDeviceFacade(Device):
 
     """
 
+    def __new__(cls, device: "qml.devices.LegacyDevice"):
+        return device if isinstance(device, cls) else super().__new__(cls)
+
     # pylint: disable=super-init-not-called
     def __init__(self, device: "qml.devices.LegacyDevice"):
+        if isinstance(device, self.__class__):
```
Comment (mudit2812) [inline pennylane/devices/legacy_facade.py:147]:
```suggestion
        if isinstance(device, type(self)):
```

File: pennylane/workflow/execution.py [hunk 7]
```diff
@@ -745,23 +603,21 @@ def cost_fn(params, x):
...
         cache = None
 
-    expand_fn = _preprocess_expand_fn(expand_fn, device, max_expansion)
-
     # changing this set of conditions causes a bunch of tests to break.
-    no_interface_boundary_required = interface is None or gradient_fn in {None, "backprop"}
+    no_interface_boundary_required = interface is None or config.gradient_method in {
+        None,
+        "backprop",
+    }
```
Comment (mudit2812) [inline pennylane/workflow/execution.py:610]: If we're no longer using `gradient_fn`, should that be deprecated as well? @trbromley

File: pennylane/workflow/qnode.py [hunk 8]
```diff
@@ -495,16 +495,8 @@ def __init__(
...
-                "The 'expansion_strategy' attribute is deprecated and will be removed  "
-                "in version 0.39. For full control over the stage to which the tape is "
-                "constructed, use the 'pennylane.workflow.construct_batch' function.",
-                qml.PennyLaneDeprecationWarning,
-            )
-
-        # Default to "gradient" to maintain default behaviour of "draw" and "specs"
-        expansion_strategy = expansion_strategy or "gradient"
+        if isinstance(device, qml.devices.LegacyDevice):
+            device = qml.devices.LegacyDeviceFacade(device)
```
Comment (mudit2812) [inline pennylane/workflow/qnode.py:499]: This is problematic because the instance check will return `True` whether `device` is of type `LegacyDeviceFacade` or `LegacyDevice` due to the new metaclass.

File: pennylane/workflow/qnode.py [hunk 9]
```diff
@@ -599,16 +590,7 @@ def __copy__(self):
...
-            return f"<QNode: device='{self.device}', interface='{self.interface}', diff_method='{self.diff_method}'>"
-
-        detail = "<QNode: wires={}, device='{}', interface='{}', diff_method='{}'>"
-        return detail.format(
-            self.device.num_wires,
-            self.device.short_name,
-            self.interface,
-            self.diff_method,
-        )
+        return f"<QNode: device='{self.device}', interface='{self.interface}', diff_method='{self.diff_method}'>"
```
Comment (mudit2812) [inline pennylane/workflow/qnode.py:593]: I think we should have separate branches here for the facade since the repr will include `LegacyDeviceFacade` in it, which I don't like.

File: pennylane/workflow/qnode.py [hunk 10]
```diff
@@ -777,25 +734,19 @@ def get_best_method(device, interface, tape=None):
             tuple[str or .TransformDispatcher, dict, .Device: Tuple containing the ``gradient_fn``,
             ``gradient_kwargs``, and the device to use when calling the execute function.
         """
+        if isinstance(device, qml.devices.LegacyDevice):
+            device = qml.devices.LegacyDeviceFacade(device)
```
Comment (mudit2812) [inline pennylane/workflow/qnode.py:738]: Another problematic instance check. I'm starting to wonder if it even makes sense to include the metaclass at this point.

**Review by mudit2812 — CHANGES_REQUESTED** [review]
Thanks @Shiro-Raven . At this point, I'm not sure if it makes sense to use the metaclass to override instance checks. It seems to be creating more problems than solving them. Other than that, it looks good mostly.

File: pennylane/workflow/qnode.py (hunk 10 above)
Comment (Shiro-Raven) [reply pennylane/workflow/qnode.py:738]: Discussed offline. Decided this is fine given the idempotency of the Facade constructor.

File: pennylane/workflow/qnode.py (hunk 8 above)
Comment (Shiro-Raven) [reply pennylane/workflow/qnode.py:499]: Resolved offline.

File: pennylane/devices/legacy_facade.py [hunk 11]
```diff
@@ -243,15 +251,17 @@ def _setup_device_config(self, execution_config):
...
         if execution_config.gradient_method == "best":
             tape = qml.tape.QuantumScript([], [])
+            if self._validate_device_method(tape):
+                config = replace(execution_config, gradient_method="device")
+                return self._setup_execution_config(config)
+
             if self._validate_backprop_method(tape):
                 config = replace(execution_config, gradient_method="backprop")
                 return self._setup_backprop_config(config)
+
```
[the remark this answers is not in the document]
Comment (Shiro-Raven) [reply pennylane/devices/legacy_facade.py:261]: To make it consistent with the priorities expected by the legacy device tests.

File: pennylane/transforms/core/transform_dispatcher.py [hunk 12]
```diff
@@ -302,6 +280,9 @@ def _device_transform(self, original_device, targs, tkwargs):
         if self._final_transform:
             raise TransformError("Device transform does not support final transforms.")
 
+        if originally_facaded := isinstance(original_device, qml.devices.LegacyDeviceFacade):
+            original_device = original_device.target_device
```
[the remark this answers is not in the document]
Comment (Shiro-Raven) [reply pennylane/transforms/core/transform_dispatcher.py:284]: Found a better solution 👌

File: pennylane/devices/legacy_facade.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/devices/legacy_facade.py:261]: Just a mistake on my side. Sorry.

File: pennylane/workflow/execution.py (hunk 7 above)
Comment (albi3ro) [reply pennylane/workflow/execution.py:610]: using gradient_fn to construct an `ExecutionConfig` if none is provided.  Also if `gradient_fn` is a gradient trasnform.

File: pennylane/workflow/jacobian_products.py [hunk 13]
```diff
@@ -401,8 +401,11 @@ def __init__(
         execution_config: Optional["qml.devices.ExecutionConfig"] = None,
```
Comment (albi3ro) [inline pennylane/workflow/jacobian_products.py:401]:
```suggestion
        execution_config: qml.devices.ExecutionConfig = qml.devices.DefaultExecutionConfig,
```

Comment (Shiro-Raven) [reply pennylane/workflow/jacobian_products.py]: That wouldn't work because of circular imports 😅

Comment (Shiro-Raven) [reply pennylane/workflow/jacobian_products.py]: I already switch the None to the default config in the body of the function

File: pennylane/_device.py [hunk 14]
```diff
@@ -103,7 +103,25 @@ class DeviceError(Exception):
...
+    checking "isinstance(dev, qml.devices.DefaultMixed)", the overridden
+    "__instancecheck__" will look behind the facade, and will evaluate instead
+    "isinstance(dev.target_device, qml.devices.DefaultMixed)"
+    """
+
+    def __instancecheck__(cls, instance):
+        if isinstance(instance, qml.devices.LegacyDeviceFacade):
+            return isinstance(instance.target_device, cls)
+
+        return super().__instancecheck__(instance)
```
Comment (albi3ro) [inline pennylane/_device.py:121]: At the same time, should be we able to have `isinstance(dev, qml.devices.Device) == True`? 

```suggestion
        if isinstance(instance, qml.devices.LegacyDeviceFacade) and isinstance(instance.target_device, cls):
            return True

        return super().__instancecheck__(instance)
```

File: pennylane/_device.py [hunk 15]
```diff
@@ -530,7 +548,7 @@ def batch_execute(self, circuits):
 
         return results
 
-    def execute_and_gradients(self, circuits, method="jacobian", **kwargs):
+    def execute_and_gradients(self, circuits, method="adjoint_jacobian", **kwargs):
```
Comment (albi3ro) [inline pennylane/_device.py:551]: I'd be rather cautious of this change.  What is this change needed for?

File: pennylane/devices/legacy_facade.py [hunk 16]
```diff
@@ -139,12 +139,19 @@ class LegacyDeviceFacade(Device):
...
     """
 
+    def __new__(cls, device: "qml.devices.LegacyDevice", *args, **kwargs):
+        return device if isinstance(device, cls) else super().__new__(cls)
+
     # pylint: disable=super-init-not-called
     def __init__(self, device: "qml.devices.LegacyDevice"):
+        if isinstance(device, type(self)):
+            return
+
```
Comment (albi3ro) [inline pennylane/devices/legacy_facade.py:149]:
```suggestion
```

File: tests/devices/test_default_mixed_autograd.py [hunk 17]
```diff
@@ -120,7 +120,7 @@ class TestDtypePreserved:
     """Test that the user-defined dtype of the device is preserved for QNode
     evaluation"""
 
-    @pytest.mark.parametrize("r_dtype", [np.float32, np.float64])
+    @pytest.mark.parametrize("r_dtype_name", ["float32", "float64"])
```
Comment (albi3ro) [inline tests/devices/test_default_mixed_autograd.py:123]: Why this change?

File: tests/interfaces/legacy_devices_integration/test_autograd_qnode_legacy.py [hunk 18]
```diff
@@ -1705,8 +1705,8 @@ def circuit(data, weights, coeffs):
 
         # test second-order derivatives
         if diff_method == "parameter-shift" and max_diff == 2:
-            with pytest.warns(UserWarning, match=r"Output seems independent of input."):
```
Comment (albi3ro) [inline tests/interfaces/legacy_devices_integration/test_autograd_qnode_legacy.py:1708]: No more warning?

File: pennylane/workflow/qnode.py
Comment (albi3ro) [inline pennylane/workflow/qnode.py:1]: We can now get rid of `QNode._tape_cached` and the following lines:
```
        cache = self.execute_kwargs.get("cache", False)
        using_custom_cache = (
            hasattr(cache, "__getitem__")
            and hasattr(cache, "__setitem__")
            and hasattr(cache, "__delitem__")
        )
        self._tape_cached = using_custom_cache and self.tape.hash in cache
```

as well.

File: pennylane/workflow/jacobian_products.py [hunk 19]
```diff
@@ -398,11 +398,14 @@ def __repr__(self):
     def __init__(
         self,
         device: Union["qml.devices.Device", "qml.Device"],
-        execution_config: Optional["qml.devices.ExecutionConfig"] = None,
+        execution_config: "qml.devices.ExecutionConfig" = None,
```
[the remark this answers is not in the document]
Comment (Shiro-Raven) [reply pennylane/workflow/jacobian_products.py:401]: Yeah exactly

File: pennylane/_device.py (hunk 15 above)
Comment (Shiro-Raven) [reply pennylane/_device.py:551]: The old default value caused several failures in the `test_jacobian_products` tests

```bash
FAILED tests/interfaces/test_jacobian_products.py::TestJacobianProductResults::test_execute_jvp_basic[None-jpc4] - AttributeError: 'DefaultQubitLegacy' object has no attribute 'jacobian'
FAILED tests/interfaces/test_jacobian_products.py::TestJacobianProductResults::test_vjp_basic[None-jpc4] - AttributeError: 'DefaultQubitLegacy' object has no attribute 'jacobian'
FAILED tests/interfaces/test_jacobian_products.py::TestJacobianProductResults::test_jacobian_basic[None-jpc4] - AttributeError: 'DefaultQubitLegacy' object has no attribute 'jacobian'
FAILED tests/interfaces/test_jacobian_products.py::TestJacobianProductResults::test_execute_jacobian_basic[None-jpc4] - AttributeError: 'DefaultQubitLegacy' object has no attribute 'jacobian'
FAILED tests/interfaces/test_jacobian_products.py::TestJacobianProductResults::test_batch_execute_jvp[None-jpc4] - AttributeError: 'DefaultQubitLegacy' object has no attribute 'jacobian'
FAILED tests/interfaces/test_jacobian_products.py::TestJacobianProductResults::test_batch_vjp[None-jpc4] - AttributeError: 'DefaultQubitLegacy' object has no attribute 'jacobian'
FAILED tests/interfaces/test_jacobian_products.py::TestJacobianProductResults::test_batch_jacobian[None-jpc4] - AttributeError: 'DefaultQubitLegacy' object has no attribute 'jacobian'
FAILED tests/interfaces/test_jacobian_products.py::TestJacobianProductResults::test_batch_execute_jacobian[None-jpc4] - AttributeError: 'DefaultQubitLegacy' object has no attribute 'jacobian'
FAILED tests/interfaces/test_jacobian_products.py::TestCachingDeviceDerivatives::test_execution_caching[jpc1] - AttributeError: 'DefaultQubitLegacy' object has no attribute 'jacobian'
FAILED tests/interfaces/test_jacobian_products.py::TestCachingDeviceDerivatives::test_cached_on_execute_and_compute_jvps[jpc1] - AttributeError: 'DefaultQubitLegacy' object has no attribute 'jacobian'
FAILED tests/interfaces/test_jacobian_products.py::TestCachingDeviceDerivatives::test_cached_on_execute_and_compute_jacobian[jpc1] - AttributeError: 'DefaultQubitLegacy' object has no attribute 'jacobian'
FAILED tests/interfaces/test_jacobian_products.py::TestCachingDeviceDerivatives::test_cached_on_vjps[jpc1] - AttributeError: 'DefaultQubitLegacy' object has no attribute 'jacobian'
```

File: tests/devices/test_default_mixed_autograd.py (hunk 17 above)
Comment (Shiro-Raven) [reply tests/devices/test_default_mixed_autograd.py:123]: I actually don't know. I think some test was failing and I thought this fixed it. Probably just a red herring. I reverted the change

File: tests/interfaces/legacy_devices_integration/test_autograd_qnode_legacy.py (hunk 18 above)
Comment (Shiro-Raven) [reply tests/interfaces/legacy_devices_integration/test_autograd_qnode_legacy.py:1708]: Nope. Is that a problem?

Comment (albi3ro) [reply tests/interfaces/legacy_devices_integration/test_autograd_qnode_legacy.py:1708]: This just seems rather suspicious.

Comment (albi3ro) [reply tests/interfaces/legacy_devices_integration/test_autograd_qnode_legacy.py:1708]: something to do with running `hamiltonian_expand`in the "inner execute" instead of outside the ml boundary. I will further investigate.

Especially since we actually get:
```
>>> qml.jacobian(qml.grad(circuit, argnum=2), argnum=2)(d, w, c)
<autograd.numpy.numpy_boxes.ArrayBox at 0x30556a600>
```

which is probably not an answer we want to return.

File: pennylane/transforms/tape_expand.py [hunk 20]
```diff
@@ -505,7 +505,9 @@ def circuit():
     1: ──H─╰Z──H─┤
 
     """
-    if isinstance(dev, qml.devices.LegacyDevice):
+    if isinstance(dev, qml.devices.LegacyDeviceFacade):
+        dev = dev.target_device
```
Comment (mudit2812) [inline pennylane/transforms/tape_expand.py:509]: Isn't this against the purpose of this PR?

File: tests/measurements/legacy/test_classical_shadow_legacy.py [hunk 21]
```diff
@@ -48,63 +43,20 @@ def circuit():
     return circuit
 
 
-def get_x_basis_circuit(wires, shots, interface="autograd", device="default.qubit.legacy"):
+def get_basis_circuit(wires, shots, basis, interface="autograd", device="default.qubit.legacy"):
     """
-    Return a QNode that prepares the |++..+> state and performs a classical shadow measurement
+    Return a QNode that prepares the |00..0> state and performs a classical shadow measurement
```
Comment (mudit2812) [inline tests/measurements/legacy/test_classical_shadow_legacy.py:48]: But isn't this test being parametrized over all Pauli bases?

File: tests/ops/qutrit/test_qutrit_parametric_ops.py [hunk 22]
```diff
@@ -465,6 +465,9 @@ def circuit(phi):
     @pytest.mark.parametrize("diff_method", diff_methods)
     def test_differentiability_broadcasted(self, op, obs, grad_fn, diff_method, tol):
         """Test that differentiation of parametrized operations with broadcasting works."""
+        if diff_method in ("finite-diff", "parameter-shift"):
+            pytest.xfail()
```
Comment (mudit2812) [inline tests/ops/qutrit/test_qutrit_parametric_ops.py:469]: Why do these diff methods need to xfail?

File: tests/pulse/test_parametrized_evolution.py [hunk 23]
```diff
@@ -544,21 +544,23 @@ def test_return_intermediate_and_complementary(self, comp, len_t):
...
+        from pennylane.devices.default_qubit_jax import DefaultQubitJax
+
         if time_interface == "jax":
             time = jnp.array(time)
         elif time_interface == "numpy":
             time = np.array(time)
         H = qml.pulse.ParametrizedHamiltonian([2], [qml.PauliX(0)])
 
-        dev = device_class(wires=1)
+        dev = {**globals(), **locals()}[device_class](wires=1)
```
Comment (mudit2812) [inline tests/pulse/test_parametrized_evolution.py:563]: I don't understand what the need for this was.

File: tests/test_debugging.py [hunk 24]
```diff
@@ -191,7 +191,10 @@ def circuit():
     @pytest.mark.parametrize("diff_method", [None, "parameter-shift"])
     def test_all_state_measurement_snapshot_pure_qubit_dev(self, dev, diff_method):
         """Test that the correct measurement snapshots are returned for different measurement types."""
-        if isinstance(dev, (qml.devices.default_mixed.DefaultMixed, qml.QutritDevice)):
+        if isinstance(
+            dev,
+            (qml.devices.default_mixed.DefaultMixed, qml.QutritDevice),
+        ):
```
Comment (mudit2812) [inline tests/test_debugging.py:197]:
```suggestion
        if isinstance(dev, (qml.devices.default_mixed.DefaultMixed, qml.QutritDevice)):
```

File: pennylane/workflow/qnode.py [hunk 25]
```diff
@@ -590,7 +589,16 @@ def __copy__(self):
 
     def __repr__(self):
         """String representation."""
-        return f"<QNode: device='{self.device}', interface='{self.interface}', diff_method='{self.diff_method}'>"
+        if not isinstance(self.device, qml.devices.LegacyDeviceFacade):
```
Comment (mudit2812) [inline pennylane/workflow/qnode.py:592]: Why are negated conditions being used? I think it's a bit more readable to change the predicate to `if isinstance(self.device, qml.devices.LegacyDeviceFacade)`.

File: pennylane/workflow/qnode.py [hunk 26]
```diff
@@ -495,7 +495,7 @@ def __init__(
                 gradient_kwargs,
             )
 
-        if isinstance(device, qml.devices.LegacyDevice):
+        if not isinstance(device, qml.devices.Device):
```
Comment (mudit2812) [inline pennylane/workflow/qnode.py:498]: Same comment here.

**Review by mudit2812 — COMMENTED** [review]
Looks pretty close to approval ready, just a few suggestions/comments

File: tests/test_debugging.py (hunk 24 above)
Comment (Shiro-Raven) [reply tests/test_debugging.py:197]: I am using the default settings and `.precommit` didn't complain 🤷

File: pennylane/transforms/tape_expand.py (hunk 20 above)
Comment (Shiro-Raven) [reply pennylane/transforms/tape_expand.py:509]: There's no way around it 🤷

File: pennylane/workflow/execution.py [hunk 27]
```diff
@@ -682,6 +537,9 @@ def cost_fn(params, x):
            [ 0.01983384, -0.97517033,  0.        ],
            [ 0.        ,  0.        , -0.95533649]])
     """
+    if not isinstance(device, qml.devices.Device):
```
[the remark this answers is not in the document]
Comment (Shiro-Raven) [reply pennylane/workflow/execution.py:540]: The facade is only applied if the user instantiates the device through `qml.device`. This check here is only a precautionary one in case the user creates the Legacy device independently through its constructor and passes it to the function.

File: pennylane/devices/tests/conftest.py [hunk 28]
```diff
@@ -41,6 +42,27 @@
...
+    warnings released are related to the legacy devices. Otherwise, it re-raises
+    any unrelated warnings"""
+
+    with warnings.catch_warnings(record=True) as recwarn:
+        warnings.simplefilter("always")
+        yield
+
+        for w in recwarn:
+            if isinstance(w, qml.PennyLaneDeprecationWarning):
+                assert "Use of 'default.qubit." in str(w.message)
```
[the remark this answers is not in the document]
Comment (Shiro-Raven) [reply pennylane/devices/tests/conftest.py:57]: So the fixtures work in a hierarchical order, and only captures warnings that are either: not captured by the test, or re-raised by the test if the warning is processed in the test function. Therefore, we should be safe. This fixture is gonna be around until the deprecation of the legacy devices is complete anyway.

File: pennylane/devices/legacy_facade.py [hunk 29]
```diff
@@ -139,12 +157,16 @@ class LegacyDeviceFacade(Device):
 
     """
 
+    def __new__(cls, device: "qml.devices.LegacyDevice", *args, **kwargs):
+        return device if isinstance(device, cls) else super().__new__(cls)
```
[the remark this answers is not in the document]
Comment (Shiro-Raven) [reply pennylane/devices/legacy_facade.py:161]: It's needed to avoid adding an accidental extra facade around an already facaded device. Just a way to ensure idempotency.

File: pennylane/devices/legacy_facade.py [hunk 30]
```diff
@@ -98,6 +103,15 @@ def legacy_device_batch_transform(tape, device):
     return _set_shots(device, tape.shots)(device.batch_transform)(tape)
 
 
+def adjoint_ops(op: qml.operation.Operator) -> bool:
+    """Specify whether or not an Operator is supported by adjoint differentiation."""
+    if isinstance(op, qml.QubitUnitary) and not qml.operation.is_trainable(op):
+        return True
```
[the remark this answers is not in the document]
Comment (Shiro-Raven) [reply pennylane/devices/legacy_facade.py:109]: @albi3ro I think you can answer this question better than I do 😅

File: pennylane/workflow/execution.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (astralcai) [reply pennylane/workflow/execution.py:540]: In that case, why not override `LegacyDevice.__new__` to always return a `LegacyDeviceFacade`?

File: pennylane/devices/legacy_facade.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (astralcai) [reply pennylane/devices/legacy_facade.py:161]: Users wouldn't be directly creating the facade, so I would think that if there's code somewhere that creates a facade around an already facaded device, it should be considered a bug.

File: pennylane/workflow/execution.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (Shiro-Raven) [reply pennylane/workflow/execution.py:540]: @albi3ro I actually think this might be quite efficient. What do you think?

File: pennylane/transforms/core/transform_dispatcher.py [hunk 31]
```diff
@@ -328,7 +306,9 @@ def original_device(self):
                 """Return the original device."""
                 return self._original_device
 
-        return TransformedDevice(original_device, self._transform)
+        new_dev = TransformedDevice(original_device, self._transform)
+
+        return new_dev
```
[the remark this answers is not in the document]
Comment (Shiro-Raven) [reply pennylane/transforms/core/transform_dispatcher.py:311]: No reason. I reverted the change

File: pennylane/devices/tests/test_measurements.py [hunk 32]
```diff
@@ -1783,15 +1783,15 @@ def circuit():
...
                 UserWarning,
-                match="Requested measurement MyMeasurement with finite shots",
-            ):
-                circuit()
-        else:
-            with pytest.raises(qml.DeviceError):
-                circuit()
+                match="MyMeasurement with finite shots; the returned state information is analytic",
+            )
+            if isinstance(dev, qml.devices.LegacyDevice)
```
[the remark this answers is not in the document]
Comment (Shiro-Raven) [reply pennylane/devices/tests/test_measurements.py:1791]: Because the facade is merely an adapter. The underlying code for executing measurements is still different and will provide different warnings/errors.

File: pennylane/workflow/execution.py [hunk 33]
```diff
@@ -785,24 +641,15 @@ def inner_execute_with_empty_jac(tapes, **_):
...
-                UserWarning,
-            )
-        tapes, post_processing = transform_program(tapes)
-    else:
-        # TODO: Remove once old device are removed
-        tapes, program_post_processing = transform_program(tapes)
-        tapes, program_pre_processing, config = _batch_transform(
-            tapes, device, config, override_shots, device_batch_transform
+    #### Executing the configured setup #####
+    if not device_batch_transform:
```
[the remark this answers is not in the document]
Comment (Shiro-Raven) [reply pennylane/workflow/execution.py:646]: Right. It's a deprecated argument, with a default of `None`. But if you check [here](https://github.com/PennyLaneAI/pennylane/blob/ad/facade-wrapper/pennylane/workflow/execution.py#L300), you can see it being defaulted to `True`.

File: pennylane/workflow/execution.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/workflow/execution.py:540]: I see metaprogramming and Python black magic like `__new__` as a tool of last resort when we have no better options.  Yes, I have used it several times in the past, but that has also made me highly skeptical of it when we have different options.  

In this case, our other option is to just add a couple `if isinstance(device, LegacyDevice): device = LegacyDeviceFacade(device)`, which is pretty minimal in terms of code.  Any python developer can understand what it's doing.  I'm not concerned with the complexity and maintainence burden of those lines.

**Review by astralcai — APPROVED** [review]
LGTM! Impressive work! 🚀

File: pennylane/devices/legacy_facade.py [hunk 34]
```diff
@@ -98,6 +103,15 @@ def legacy_device_batch_transform(tape, device):
...
+    return not isinstance(op, MidMeasureMP) and (
```
Comment (mudit2812) [inline pennylane/devices/legacy_facade.py:110]: In anticipation of #5868 

```suggestion
    return not isinstance(op, (MidMeasureMP, qml.ops.Conditional)) and (
```

Comment (mudit2812) [reply pennylane/devices/legacy_facade.py:110]: Wait ignore this. This should be resolved in the same PR, not here.

**Review by albi3ro — APPROVED** [review]
Fantastic job on a rather difficult piece of work.  I am excited to have finally reached this threshold in a very, very long process. This is going to make development of the workflow much easier going forward, and unlock a lot of technical improvements and code quality possibilities.
